### PR TITLE
WIP on Railtie

### PR DESCRIPTION
**Pull request by danmayer**
I think this might be ready.

* note I am not sure if the background reporting released in 3.0.1 was usable @kbaum see that I had to add the force report to bypass the background check... the issue I hit was that even when the background job tried to report coverage it was a no-op because the first line just returned... Perhaps I missed something?
* so far this all seems good for getting initial data when the app boots, and for the background 30-second reporting

### CODE REVIEW

File: lib/coverband/collectors/coverage.rb [hunk 1]
```diff
@@ -31,7 +31,7 @@ def reset_instance
       end
 
       def report_coverage(force_report = false)
-        return Background.start if @background_reporting_enabled
+        return Background.start if @background_reporting_enabled && !force_report
```
Comment (kbaum) [inline lib/coverband/collectors/coverage.rb]: @danmayer definitely needed. I don't think I tested the actual report generation enough and was more concerned with how and when the background thread was running. Sorry about that. Should I backport your fix for a 3.02 release?

File: lib/coverband/utils/railtie.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module Coverband
+  class Railtie < Rails::Railtie
+    # Coverband needs to be setup before any of the initializers
+    # to capture usage of things loaded by them
+    # if one uses before_eager_load as I did previously
+    # any files that get loaded as part of railties will have no coverage
+    config.before_initialize do
+      Coverband.configure
```
Comment (kbaum) [inline lib/coverband/utils/railtie.rb]: This line currently fails if there is no coverband.rb. What do you think of changing this and just going with sensible defaults?

File: lib/coverband/collectors/coverage.rb (hunk 1 above)
Comment (kbaum) [reply lib/coverband/collectors/coverage.rb]: Should we consider kicking off Background.start from the middleware instead of here? Then we don't have to worry about the force_report flag.

File: lib/coverband/utils/railtie.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    initializer 'coverband.configure' do |app|
+      app.middleware.use Coverband::Middleware
+    end
+
+    config.after_initialize do
+      Coverband::Collectors::Coverage.instance.report_coverage(true)
+    end
+
+    rake_tasks do
+      load 'coverband/utils/tasks.rb'
```
Comment (kbaum) [inline lib/coverband/utils/railtie.rb]: Nice! I didn't realize there was a hook for rake tasks!

File: lib/coverband/version.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module Coverband
-  VERSION = '3.0.1'
+  VERSION = '3.0.2.alpha'
```
Comment (kbaum) [inline lib/coverband/version.rb]: Are we targeting this for 3.0.2 or 4.x?

File: lib/coverband/utils/railtie.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    # if one uses before_eager_load as I did previously
+    # any files that get loaded as part of railties will have no coverage
+    config.before_initialize do
+      Coverband.configure
+      Coverband.configuration.logger&.debug('Railtie setting up Coverband')
+      Coverband.start
+    end
+
+    initializer 'coverband.configure' do |app|
+      app.middleware.use Coverband::Middleware
```
Comment (kbaum) [inline lib/coverband/utils/railtie.rb]: If we are targeting 3.0.2, then we might need to consider this line adding 2 middlewares on accident.

File: lib/coverband/collectors/coverage.rb (hunk 1 above)
Comment (danmayer) [reply lib/coverband/collectors/coverage.rb]: yeah that makes sense, perhaps back to the idea of 2 distinct middlewares... one just to start background and the other for middleware reporting?

I am wondering if we could have middleware the removes itself after it kicks off background start so we don't have it adding any overhead after that starts.

File: lib/coverband/utils/railtie.rb (hunk 2 above)
Comment (danmayer) [reply lib/coverband/utils/railtie.rb]: Yeah I was thinking of trying to work towards the zero config side of things. I will do a pass and see if anything is really needed for basic running out of the gate. I am assuming for zero config we look for standard envs like REDIS_URL. We probably would want to give better errors if we can't detect anything.

File: lib/coverband/version.rb (hunk 4 above)
Comment (danmayer) [reply lib/coverband/version.rb]: I wasn't sure yet but needed to pick something to test locally, I didn't release the alpha gem. I think to really get full benefits we will want to do some breaking changes so I am going to move it to 4.0.0

File: lib/coverband/collectors/coverage.rb (hunk 1 above)
Comment (danmayer) [reply lib/coverband/collectors/coverage.rb]: I think we still need force report to bypast the reporting frequency...

I guess I wouldn't worry about a 3.0.2 release since we didn't have this enabled by default or fullly documented I don't think it is impacting anymore if we get a bug report we can point them towards this. We might get 4.0.0 done before it is an issue.

File: lib/coverband/utils/railtie.rb (hunk 2 above)
Comment (danmayer) [reply lib/coverband/utils/railtie.rb]: see latest commits, I think it would cover basic assumptions we could probably make it a bit more resilent but feels like a reasonable first pass.

File: lib/coverband/version.rb (hunk 4 above)
Comment (danmayer) [reply lib/coverband/version.rb]: moved to 4.0.x in latest commits

File: lib/coverband/collectors/coverage.rb (hunk 1 above)
Comment (kbaum) [reply lib/coverband/collectors/coverage.rb]: Ahhh.. totally didn't think about reporting frequency. Removing all of these configuration options will make the logic much easier to reason about. Does it make sense to target removing reporting_frequency altogether considering we are now reporting from the background thread?

Agreed on not releasing a 3.0.2. Let's push for 4.0.0.

File: lib/coverband/configuration.rb [hunk 6]
```diff
@@ -56,6 +61,12 @@ def s3_secret_access_key
...
     def store
       return @store if @store
+      @store = if ENV['REDIS_URL']
+                 Coverband::Adapters::RedisStore.new(Redis.new(url: ENV['REDIS_URL']))
+               else
+                 Coverband::Adapters::RedisStore.new(Redis.new)
+               end
+      return @store if @store
       raise 'no valid store configured'
     end
```
Comment (kbaum) [inline lib/coverband/configuration.rb]: Was thinking it might make sense to try COVERBAND_REDIS_URL then REDIS_URL.

```ruby
def store
  @store ||= Coverband::Adapters::RedisStore.new(Redis.new(url: redis_url))
end

def redis_url
  ENV['COVERBAND_REDIS_URL'] || ENV['REDIS_URL']
end
```

This will give an easy way to specific an alternate redis instance in production.

Comment (danmayer) [reply lib/coverband/configuration.rb]: Good call adding this... changed it up a bit to make sure we pull in redis_ttl and redis_namespace as well

Comment (danmayer) [reply lib/coverband/configuration.rb]: ok see latest commit I think this is resolved

File: lib/coverband/collectors/coverage.rb (hunk 1 above)
Comment (danmayer) [reply lib/coverband/collectors/coverage.rb]: ok take a look at latest commits @kbaum I refectored this a bit and moved background start to the middleware

**Comment by danmayer** [comment]
updated to get all tests passing

File: lib/coverband/configuration.rb [hunk 7]
```diff
@@ -55,19 +60,38 @@ def s3_secret_access_key
...
       elsif store.is_a?(String)
         @store = Coverband::Adapters::FileStore.new(store)
       end
     end
+
+    private
+
+    def coverband_redis
+      # Redis doesn't default to correct local host if you pass url: nil
+      redis_url ? Redis.new(url: redis_url) : Redis.new
```
Comment (kbaum) [inline lib/coverband/configuration.rb]: Not that important but Redis.new(url: nil) seems to work for me:
```
[4] pry(main)> Redis.new(url: nil)
#<Redis client v4.0.3 for redis://127.0.0.1:6379/0>
[5] pry(main)>
```

File: lib/coverband/configuration.rb [hunk 8]
```diff
@@ -55,19 +60,38 @@ def s3_secret_access_key
...
+    def redis_url
+      ENV['COVERBAND_REDIS_URL'] || ENV['REDIS_URL']
+    end
+
+    def redis_store_options
+      { ttl: Coverband.configuration.redis_ttl,
+        redis_namespace: Coverband.configuration.redis_namespace }
+    end
+
+    def redis_options
```
Comment (kbaum) [inline lib/coverband/configuration.rb]: Are both redis_options and redis_store_options methods needed?

File: lib/coverband/integrations/middleware.rb [hunk 9]
```diff
@@ -9,7 +9,11 @@ def initialize(app)
     def call(env)
       @app.call(env)
     ensure
-      Coverband::Collectors::Coverage.instance.report_coverage
+      if Coverband.configuration.background_reporting_enabled
+        Coverband::Background.start
+      else
+        Coverband::Collectors::Coverage.instance.report_coverage
```
Comment (kbaum) [inline lib/coverband/integrations/middleware.rb]: Do we even want to keep this option around?

File: lib/coverband/configuration.rb (hunk 7 above)
Comment (danmayer) [reply lib/coverband/configuration.rb]: weird I swear I had had an issue with that in the past...

File: lib/coverband/configuration.rb (hunk 8 above)
Comment (danmayer) [reply lib/coverband/configuration.rb]: oops no I meat to get rid of the redis_options doing that now...

File: lib/coverband/configuration.rb (hunk 7 above)
Comment (danmayer) [reply lib/coverband/configuration.rb]: fixed

File: lib/coverband/integrations/middleware.rb (hunk 9 above)
Comment (danmayer) [reply lib/coverband/integrations/middleware.rb]: yeah see other comments, I'd rather not get rid of the middleware flow until we ensure on all setups the background thread works well. Feels like it would be good to have both as an option until we get a bit more real-world testing and feedback on the new background flow.

Comment (kbaum) [reply lib/coverband/integrations/middleware.rb]: Agreed.

File: README.md [hunk 10]
```diff
@@ -96,86 +87,48 @@ $ gem install coverband
...
-After you have the gem, you will want to setup the needed integration:
+After you have the gem, you may need to configure additional options:
 
-1. Coverband Config File Setup
-2. Configure Rake
-3. Insert Coverband Rack Middleware
+### 1. Verify Rake
 
-### 1. Coverband Config File Setup
+After installing the gem in Rails you should have these Rake tasks
```
Comment (kbaum) [inline README.md]: I think it's important that the readme gives a feeling of a drop in gem.. especially for rails. Was thinking the basic installation would be communicated near the top. Those who are interested in additional configuration options have to do a bit more digging in. Something like:

## General installation
### Rails
### Sinatra
## Report Generation
### Rake Task
### Viewing the Report in App
#### Rails
#### Sinatra
## Additional Configuration

Comment (kbaum) [reply README.md]: Just wanted to add.. happy to take a crack at it if you like the basic idea.

File: lib/coverband/configuration.rb [hunk 11]
```diff
@@ -55,19 +59,28 @@ def s3_secret_access_key
     end
 
     def store
-      return @store if @store
-      raise 'no valid store configured'
+      @store ||= Coverband::Adapters::RedisStore.new(Redis.new(url: redis_url), redis_store_options)
     end
 
     def store=(store)
       if store.is_a?(Coverband::Adapters::Base)
```
Comment (kbaum) [inline lib/coverband/configuration.rb]: What do you think of removing the option of passing in a redis instance here?  Make this config option only take a store. Would really clean up this method and it seems pretty straight forward.

Comment (danmayer) [reply lib/coverband/configuration.rb]: yeah that was for old backward compatibility now is a good time to remove. Doing that. Thanks

File: lib/coverband/configuration.rb [hunk 12]
```diff
@@ -55,19 +59,28 @@ def s3_secret_access_key
...
       end
     end
+
+    private
+
+    def redis_url
+      ENV['COVERBAND_REDIS_URL'] || ENV['REDIS_URL']
+    end
+
+    def redis_store_options
```
[the remark this answers is not in the document]
Comment (danmayer) [reply lib/coverband/configuration.rb]: Not having namespace or ttl as part of the internal knowledge of the system would be nice, but would be very different and remove some specific features. Naming sharing a single Redis for multiple environments or applications...

Basically, redis-rb doesn't support ttl or namespace at a connection level at all. https://github.com/redis/redis-rb

TTL you simply have to call per key which is what Coverband does unless I am missing something https://github.com/danmayer/coverband/blob/feature/railtie/lib/coverband/adapters/redis_store.rb#L32

---

Namespace... It is kind of an annoyingly controversial issue...

* https://github.com/redis/redis-rb/issues/222

This caused various folks to declare the redis-namespace gem a hack, "which should never be used". While others want to use a single Redis without polluting the top level namespace... https://www.mikeperham.com/2017/04/10/migrating-from-redis-namespace/

Basically, I don't want to have to add in the Redis namespace gem... and what Coverband does it also a simple naming hack... https://github.com/danmayer/coverband/blob/feature/railtie/lib/coverband/adapters/redis_store.rb#L27

Reasons to keep it, one can have staging(s) and production share a Redis using this feature, which is the original reason it was requested in the past. Reasons to get rid of it... is that it is just a naming convention, we could push folks asking them to Redis dbs as Sidekiq did, but seems it has performance issues.

I think it is worth supporting multiple apps and environments via these options.

File: lib/coverband/configuration.rb (hunk 11 above)
Comment (danmayer) [reply lib/coverband/configuration.rb]: done

File: lib/coverband/configuration.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (kbaum) [reply lib/coverband/configuration.rb]: Totally makes sense. Thanks for the explanation! Guess I’m just giving all the configuration options a second look while we have the chance.

File: README.md (hunk 10 above)
Comment (danmayer) [reply README.md]: Thanks, I took another pass moving some of the more advanced stuff to the end and trying to simplify the earlier parts of the Readme. Take a look and let me know if that looks better @kbaum

File: README.md [hunk 13]
```diff
@@ -5,190 +5,126 @@ Build Status: [![Build Status](https://travis-ci.org/danmayer/coverband.svg?bran
...
+  # ... lots of routes
+  mount Coverband::Reporters::Web.new, at: '/coverage'
 end
 ```
 
-#### For Sinatra apps
+__NOTE__: ADD PASSWORD PROTECTION OR YOU CAN EXPOSE ALL YOUR SOURCE CODE
 
-For the best coverage you want this loaded as early as possible. I have been putting it directly in my `config.ru` but you could use an initializer, though you may end up missing some boot up coverage.
+It is easy to add some basic protect around the coverage data, below shows how you can use devise or basic auth, by adding a bit of code to your `config/routes.rb` file.
```
Comment (kbaum) [inline README.md]: "basic protection"

File: README.md [hunk 14]
```diff
@@ -199,97 +135,107 @@ run ActionController::Dispatcher.new
...
 
-### Example apps
+ Below is an example config file for a Rails 5 app:
 
-- [Rails 5.2.x App](https://github.com/danmayer/coverage_demo)
-- [Sinatra app](https://github.com/danmayer/churn-site)
-- [Non Rack Ruby app](https://github.com/danmayer/coverband_examples)
+```ruby
+#config/coverband.rb
+Coverband.configure do |config|
```
Comment (kbaum) [inline README.md]: Think we should mention that we can set the environment variables or configure through coverband with config.s3_access_key_id and config.s3_secret_access_key?

File: README.md [hunk 15]
```diff
@@ -305,6 +251,7 @@ def after_perform(*args)
 end
 ```
 
+(no need to do this for Rails, but a Rack app maybe)
```
Comment (kbaum) [inline README.md]: This part may be confusing. For sidekiq and no rails, the default background thread removes the need for custom sidekiq middleware. The only custom code needed should be Coverband.configure and Coverband.start. I am thinking we should probably even remove the need for this. Maybe:

```ruby
if defined?(Sidekiq) && !defined?(Rails)
  Sidekiq.configure_server do
    Coverband.configure
     Coverband.start
  end
end
```

While we are on the topic of configure and start, would it be possible to just call these methods on require of coverband.rb and remove this requirement altogether?

File: README.md (hunk 13 above)
Comment (danmayer) [reply README.md]: fixed

File: README.md (hunk 14 above)
Comment (danmayer) [reply README.md]: good call updated docs around this as well

File: README.md (hunk 15 above)
Comment (danmayer) [reply README.md]: Yeah so I was thinking for Ruby but not Rails Sidekiq apps... I think your idea of automatically calling configure and start even outside of the railtie is an interesting one. Let me run some tests on some plain ruby scripts and see if I have any Sinatra apps to test with, It seems like a good idea, but I would want to test it out a bit on some samples first.

File: README.md [hunk 16]
```diff
@@ -337,6 +284,8 @@ Coverband::Collectors::Coverage.instance.report_coverage
 
 ### Manual Configuration (for cron jobs / Raketasks)
```
[the remark this answers is not in the document]
Comment (danmayer) [reply README.md]: Same as above, I am going to test this a bit if it works well I will drop these doc sections and replace with an explanation.

File: README.md (hunk 15 above)
Comment (danmayer) [reply README.md]: ok, take a look at the proof of concept of this idea...

added on latest commit to this branch... https://github.com/danmayer/coverband/commit/3240a161f8739150ed18473647284914fffaaf97

An example script which collects its own runtime data via this method
https://github.com/danmayer/coverband_demo/pull/2/files#diff-7ce9c6873a144d3d0eb1ca7010c03a86R1

The Coverband demo Rails app still works as well.

Comment (danmayer) [reply README.md]: Let me know if you see any issues with this, if not I will try to do a bit more testing and clean up docs @kbaum

File: README.md [hunk 17]
```diff
@@ -5,190 +5,126 @@ Build Status: [![Build Status](https://travis-ci.org/danmayer/coverband.svg?bran
...
 * Low performance overhead
-* Various controls from sampling, data collection rate, etc to further control performance
-* Ignore directories to avoid overhead data collection on vendor, lib, etc.
+* Very simple setup and configuration
+* Out of the box support for all standard code execution paths (web, cron, background jobs, rake tasks, etc)
+* Easy to understand actionable insights from the report
 * Development mode, offers deep insight of code usage details (number of LOC execution during single request, etc) during development.
-* Easy setup for any Ruby Rack based web framework (Rails, Sinatra, etc)
-* Allows for integration with any other Ruby application flows (background jobs, crons, scripts)
+* Mountable web interface to easily share reports
```
Comment (kbaum) [inline README.md]: The features section looks great!

File: README.md [hunk 18]
```diff
@@ -5,190 +5,126 @@ Build Status: [![Build Status](https://travis-ci.org/danmayer/coverband.svg?bran
...
-## Example Output
-
-Since Coverband is [Simplecov](https://github.com/colszowka/simplecov) output compatible it should work with any of the `SimpleCov::Formatter`'s available. The output below is produced using the default Simplecov HTML formatter.
-
-Index Page
-![image](https://raw.github.com/danmayer/coverband/master/docs/coverband_index.png)
-
-Details on a example Sinatra app
-![image](https://raw.github.com/danmayer/coverband/master/docs/coverband_details.png)
+Take Coverband for a spin on the live Heroku deployed [Coverband Demo](https://coverband-demo.herokuapp.com/). The [full source code for the demo](https://github.com/danmayer/coverband_demo) is available to help with installation, configuration, and understanding of basic usage.
```
Comment (kbaum) [inline README.md]: I think the demo section should be moved lower. When i look at a readme, i want to see what the gem is and then how to install it.

File: README.md [hunk 19]
```diff
@@ -5,190 +5,126 @@ Build Status: [![Build Status](https://travis-ci.org/danmayer/coverband.svg?bran
...
 ## Gem Installation
 
-Add this line to your application's Gemfile:
+Add this line to your application's `Gemfile`, remember to `bundle install` after updating:
 
 ```bash
 gem 'coverband'
 ```
 
-And then execute:
```
Comment (kbaum) [inline README.md]: Love deleting this part.

File: README.md [hunk 20]
```diff
@@ -199,97 +135,116 @@ run ActionController::Dispatcher.new
...
+* <a href="#installation">Install Coverband</a>
+* Start your app and hit a few endpoints
+* Validate data collection and code coverage with  `rake coverband:coverage`
+* If you see app startup and recent visits showing, setup is correct
+* I generally configure the mountable web endpoint to [view the data via the web-app](https://github.com/danmayer/coverband#viewing--hosting-s3-coverband-results-in-app)
+* After Coverband has been verified to be working on production, I let it run for a few weeks.
+* Then I view the report and start submitting PRs for the team to review delete large related sets of code that no longer are in use.
 
-bundle install
+# Advanced Usage
```
Comment (kbaum) [inline README.md]: I think we can delete this header?

File: README.md [hunk 21]
```diff
@@ -400,9 +358,6 @@ By adding any files above you will get reporting on those files as part of your
 
 By default Coverband has assumed you are trying to track your application code usage not all the supporting framework and library (Gems) code usage. There has been some good points and reasons folks want to track library usage, for example to find out which Gems they aren't actually using in production. See some of the discussion on [issue 21](https://github.com/danmayer/coverband/issues/21).
 
-* Using the Tracepoint Collector
-   * just make sure to set `Coverband.configuration.include_gems = true` in your config options
-   * note by tracking all the additional usage this can have significant performance impacts of the Tracepoint implementation
 * Using the Coverage Collector
```
Comment (kbaum) [inline README.md]: Since there is only 1 kind of collector now, we might want to reword this to be more informational. Like, "How it works" or something like that. Perhaps also moving to a wiki page?

**Comment by kbaum** [comment]
The README is really looking great. If we remove the need for calling configure and start, it will be even simpler

RE: automatically calling `Coverband.configure` and `Coverband.start` on require of the gem. Can't think of any downside to calling configure early. Coverband.start will cause the coverage to be started much earlier and possibly before the require of many other gems. This means that ruby Coverage will be tracking more lines of code. Not sure how much this impacts performance if at all but worth thinking about.

File: README.md (hunk 18 above)
Comment (danmayer) [reply README.md]: moved down.

File: README.md [hunk 22]
```diff
@@ -5,190 +5,126 @@ Build Status: [![Build Status](https://travis-ci.org/danmayer/coverband.svg?bran
...
-require 'coverband/utils/tasks'
-```
-This should give you access to a number of Coverband tasks
+## Rake Tasks
+
+After installing the gem in Rails you should have these Rake tasks
 
 ```bash
 rake -T coverband
 rake coverband:clear         # reset coverband coverage data
```
[the remark this answers is not in the document]
Comment (danmayer) [reply README.md]: I updated the comment, but I think it is helpful for folks when trying to setup or do development

` reset Coverband coverage data, helpful for development, debugging, etc`

File: README.md (hunk 20 above)
Comment (danmayer) [reply README.md]: done

File: README.md (hunk 21 above)
Comment (danmayer) [reply README.md]: updated and cleaned up a bit. Yeah, I do think some of these might be good to move to the Wiki page. I think I will do that outside of this PR as it is getting a bit big ;)

**Comment by danmayer** [comment]
Yeah the only downside is we would, in theory, start the collector much earlier during code loading... so Coverage will be collecting data on many more Gems and Rails internals itself, while the Railtie would mostly just collect for App code... I think I will do some perf testing now to see if it seems significant enough care about, I am hoping not.

File: README.md (hunk 15 above)
Comment (danmayer) [reply README.md]: removed all these docs as we have autostart on require working now

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (danmayer) [reply README.md]: removed

**Comment by danmayer** [comment]
screen shot 2018-11-28 at 8 27 05 pm

Railtie benchmark on the left and the Autostart on the right... mean request time was exactly the same for both methods

**Comment by danmayer** [comment]
Alright, anything else we think should be in 4.0.0 release @kbaum or just some more perf checks and alpha testing.

File: lib/coverband/collectors/coverage.rb [hunk 23]
```diff
@@ -17,27 +17,25 @@ def self.instance
...
         self
       end
 
       def report_coverage(force_report = false)
-        return Background.start if @background_reporting_enabled
         return if !ready_to_report? && !force_report
-
         unless @store
           @logger.debug 'no store set, no-op'
           return
```
Comment (kbaum) [inline lib/coverband/collectors/coverage.rb]: Do we want to raise an error here? Could be confusing why nothing is working. Another approach could be to throw an error on Coverband.configure. That way we never even get here.

File: lib/coverband/collectors/coverage.rb [hunk 24]
```diff
@@ -74,14 +71,12 @@ def ready_to_report?
       end
 
       def get_new_coverage_results
-        coverage_results = nil
-        @semaphore.synchronize { coverage_results = new_coverage(::Coverage.peek_result.dup) }
-        coverage_results
+        @semaphore.synchronize { new_coverage(::Coverage.peek_result.dup) }
```
Comment (kbaum) [inline lib/coverband/collectors/coverage.rb]: Minor note but when we are using background thread, we should be able to avoid the `@semaphore` since there should always be 1 thread.

File: lib/coverband/collectors/coverage.rb (hunk 23 above)
Comment (kbaum) [reply lib/coverband/collectors/coverage.rb]: One other thought, is a nil @store even possible? 
https://github.com/danmayer/coverband/pull/146/files#diff-1f4b05f150269d04438736256c27b0a8R62

**Comment by kbaum** [comment]
Let's 🚢 alpha!

One other note @dnasseri and I started integrating 4.0 into our project yesterday and it is looking good so far. Probably release it into our app today or tomorrow.

File: lib/coverband/collectors/coverage.rb (hunk 24 above)
Comment (danmayer) [reply lib/coverband/collectors/coverage.rb]: yep thought about it but since I want to keep middleware option around for now we need to keep it

File: lib/coverband/collectors/coverage.rb (hunk 23 above)
Comment (danmayer) [reply lib/coverband/collectors/coverage.rb]: Yeah I had thought a bit about that... I guess it would be nice to push this into configure... basically if you don't set a store and the default one fails for some reason raise an error. 

The issue would be if the error only happens in prod it could be a scary experience... I have tried to make sure Coverband is always safe and doesn't hurt prod when it isn't working... Thoughts?

Comment (kbaum) [reply lib/coverband/collectors/coverage.rb]: Looks like we are already rescuing all errors anyway?

https://github.com/danmayer/coverband/pull/146/files#diff-181f764c69748a3ceb0cc64e4a37ebaeR43

Comment (danmayer) [reply lib/coverband/collectors/coverage.rb]: yeah true... I change to a raise which gets caught and logged, I don't think configure would get caught since that would happen outside of this rescue. Anyways should make it a bit more obvious something is wrong.

**Comment by danmayer** [comment]
OK 4.0.0.alpha is live for testing https://coverband-demo.herokuapp.com/coverage?

File: lib/coverband/collectors/coverage.rb [hunk 25]
```diff
@@ -31,10 +31,7 @@ def reset_instance
 
       def report_coverage(force_report = false)
         return if !ready_to_report? && !force_report
-        unless @store
-          @logger.debug 'no store set, no-op'
-          return
-        end
+        raise 'no Coverband store set' unless @store
```
Comment (kbaum) [inline lib/coverband/collectors/coverage.rb]: This is great because it will make it so much easier to debug people's issues.

**Comment by danmayer** [comment]
I have 4.0.0 running on a real app on staging and will get it to production tomorrow

**Comment by danmayer** [comment]
All has worked well for me @kbaum think we are good to merge and release? Or anything else we should check?

**Comment by dnasseri** [comment]
@kbaum @danmayer I've been running 4.0.0.alpha in production for two days now and everything is looking great! Nice work to both of you!

**Comment by kbaum** [comment]
@danmayer - let's merge and release!

**Comment by danmayer** [comment]
OK 4.0.0 is live!
